src/cephadm/cephadmlib/daemons/smb.py

```diff
@@ -0,0 +1,437 @@
+    def name(self) -> str:
+        return 'winbindd'
+
+    def args(self) -> List[str]:
+        return super().args() + ['run', 'winbindd']
+
+
+class ConfigInitContainer(SambaContainerCommon):
+    def name(self) -> str:
+        return 'config'
+
+    def args(self) -> List[str]:
+        return super().args() + ['init']
+
+
+class MustJoinContainer(SambaContainerCommon):
+    def name(self) -> str:
+        return 'mustjoin'
+
+    def args(self) -> List[str]:
+        args = super().args() + ['must-join']
+        for join_src in self.cfg.join_sources:
+            args.append(f'-j{join_src}')
+        return args
+
+    def container_args(self) -> List[str]:
+        cargs = []
+        for dns in self.cfg.custom_dns:
+            cargs.append(f'--dns={dns}')
+        return cargs
+
+
+class ConfigWatchContainer(SambaContainerCommon):
+    def name(self) -> str:
+        return 'configwatch'
+
+    def args(self) -> List[str]:
+        return super().args() + ['update-config', '--watch']
+
+
+class ContainerLayout:
+    init_containers: List[SambaContainerCommon]
+    primary: SambaContainerCommon
+    supplemental: List[SambaContainerCommon]
+
+    def __init__(
+        self,
+        init_containers: List[SambaContainerCommon],
+        primary: SambaContainerCommon,
+        supplemental: List[SambaContainerCommon],
+    ) -> None:
+        self.init_containers = init_containers
+        self.primary = primary
+        self.supplemental = supplemental
+
+
+@daemon_form.register
+class SMB(ContainerDaemonForm):
+    """Provides a form for SMB containers."""
+
+    daemon_type = 'smb'
+    default_image = DEFAULT_SMB_IMAGE
+
+    @classmethod
+    def for_daemon_type(cls, daemon_type: str) -> bool:
+        return cls.daemon_type == daemon_type
+
+    def __init__(self, ctx: CephadmContext, ident: DaemonIdentity):
+        assert ident.daemon_type == self.daemon_type
+        self._identity = ident
+        self._instance_cfg: Optional[Config] = None
+        self._files: Dict[str, str] = {}
+        self._raw_configs: Dict[str, Any] = context_getters.fetch_configs(ctx)
+        self._config_keyring = context_getters.get_config_and_keyring(ctx)
+        self._cached_layout: Optional[ContainerLayout] = None
+        self.smb_port = 445
+        logger.debug('Created SMB ContainerDaemonForm instance')
+
+    def validate(self) -> None:
+        if self._instance_cfg is not None:
+            return
+
+        configs = self._raw_configs
+        instance_id = configs.get('cluster_id', '')
+        source_config = configs.get('config_uri', '')
+        join_sources = configs.get('join_sources', [])
+        custom_dns = configs.get('custom_dns', [])
+        instance_features = configs.get('features', '').split()
+        files = data_utils.dict_get(configs, 'files', {})
+        ceph_config_entity = configs.get('config_auth_entity', '')
+
+        if not instance_id:
+            raise Error('invalid instance (cluster) id')
```

Comment:
is the instance id the ceph cluster FSID or something specific to SMB?

Reply:
No, it's a logical ID for one or more smbd + helper containers running on the same config and sharing the same resources. It's more like `cluster_id` in nfs. I have issues calling it `cluster` sometimes because it doesn't have to be clustered so I flip-flop back and forth on what to call it. 

I probably should admit defeat and just call it cluster_id everywhere for consistency.